src/lock/optiql.cpp

```diff
@@ -194,11 +231,33 @@ OptiQL::LockX()  //
 
 void
 OptiQL::UnlockX(  //
-    const uint64_t qid)
+    const uint64_t qid,
+    const uint64_t ver)
 {
   auto *qnode = &(_qnodes[qid]);
 
-  throw std::runtime_error{"not implemented yet"};
+  auto *next_ptr = qnode->next.load(kAcquire);
+  if (next_ptr == nullptr) {  // this is the tail node
+    auto cur = lock_.load(kRelaxed);
+    while (((cur & kQIDMask) >> kQIDShift) == qid) {
+      if (lock_.compare_exchange_weak(cur, ver, kRelease, kRelaxed)) {
+        RetainQID(qid);
+        return;
+      }
+      CPP_UTILITY_SPINLOCK_HINT
+    }
+  }
+
+  // enable opportunistic read
+  lock_.fetch_or(kOPReadLock | ver, kRelaxed);
+  while (true) {  // wait until successor fills in its next field
+    next_ptr = qnode->next.load(kAcquire);
+    if (next_ptr) break;
+    CPP_UTILITY_SPINLOCK_HINT
+  }
+
+  next_ptr->ver.store(ver + kVersionUpdate, kRelease);
```

Comment:
前述のコメントの波及でそもそもが変わりそうですが，ここもバージョンの値は意味を持たないので，ロックフラグの立っていない適当な値を渡せばOKです．また，OptiQLにおいてフェンスは共有領域（ `lock_` ）を起点に設定すれば十分なので，こっちはフェンスなしでOKです．

```suggestion
  next_ptr->ver.store(kNoLocks, kRelaxed);
```